doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,388 @@
+# Schema documentation for Avro-formatted sinks
+
+Associated issue: [#21557]
+
+## Context
+
+Avro schemas allow users to annotate records, record fields, and enums with
+documentation. Note the `doc` fields in the following example Avro schema:
+
+```json
+{
+  "type": "record",
+  "name": "row",
+  "doc": "represents a row in some materialized view",
+  "fields" : [
+    {"name": "c1", "type": "string", "doc": "this is the meaning of column a"},
+    {"name": "c2", "type": "int"}
+  ]
+}
+```
+
+When using an Avro-formatted sink, Materialize generates the Avro schema based
+on the shape of the underlying relation. For example, given the following
+sink:
+
+```sql
+CREATE MATERIALIZED VIEW v AS SELECT 'text' AS c1, 'int'::integer as c2;
+CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ...;
+```
+
+Materialize will automatically generate the following Avro schema and publish it
+to the schema registry:
+
+```json
+{
+  "type": "record",
+  "name": "envelope",
+  "fields" : [
+    {"name": "c1", "type": "string"},
+    {"name": "c2", "type": "int"}
+  ]
+}
+```
+
+Today, there is no way to instruct Materialize to include `doc` fields in the
+generated Avro schema. This poses a challenge for organizations with
+requirements on the use of documentation—e.g., a policy that requires that every
+field in every schema published to the organization's schema registry has
+documentation.
+
+## Goals
+
+Must haves:
+
+  * Enable users with organizational policies on the use of `doc` fields in Avro
+    schemas to use Avro-formatted sinks.
+
+Nice to haves:
+
+  * Consistency with Materialize's existing approach to generating Avro schemas
+    from SQL relations.
+
+  * Good ergonomics.
+
+## Non-goals
+
+None yet.
+
+## Detailed description
+
+### Design
+
+#### SQL
+
+We'll extend the `FORMAT AVRO` clause with a new `[KEY|VALUE] DOC ON
+[TYPE|COLUMN]` option. For example:
+
+```sql
+CREATE SINK ...
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ... (
+    DOC ON COLUMN <column-specifier> = '<documentation string>',
+    KEY DOC ON TYPE <type-specifier> = '<documentation string>',
+    VALUE DOC ON COLUMN <column-specifier> = '<documentation string>',
+    ...
+)
+WITH (...)
+```
+
+The option will be located in a new parenthesized option list immediately
+following the existing `FORMAT AVRO` clause. After this change, `[KEY|VALUE] DOC
+ON [TYPE|COLUMN]` will be the only supported option, but we expect to add
+further options to control the Avro schema in the future.
+
+Aside: the current design of the `USING CONFLUENT SCHEMA REGISTRY` option is
+historical and predates the introduction of the standardized option list. In the
+future, we may also move the `USING CONFLUENT SCHEMA REGISTRY` into the option
+list, where it can be freely intermixed with other options, e.g.:
+
+```sql
+FORMAT AVRO (
+    DOC ON ...,
+    CONFLUENT SCHEMA REGISTRY = '...',
+    KEY DOC ON ...,
+)
+```
```

Comment:
These format options are specifically for the published generated schema. As such, should they instead be part of existing [CSR Connection options](https://materialize.com/docs/sql/create-sink/kafka/#csr-connection-options) like `AVRO KEY FULLNAME`?

Reply:
So, something like

```
CREATE SINK ... FROM ...
FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn 
(
AVRO KEY FULLNAME = ..., 
DOC ON ...
KEY DOC ON ...
)
ENVELOPE ...
```

Reply:
Or later if we are refactoring the `USING CONFLUENT SCHEMA REGISTRY` to the option list, then to be consistent `AVRO KEY FULLNAME` should be part of the `FORMAT AVRO` options as well?

Reply:
Oh, sorry, yes, that's what I meant. I guess the design document should not have said "this will be the only option" but "these options will live alongside the existing `AVRO KEY FULLNAME` and `AVRO VALUE FULLNAME` options."

Reply:
Done!